File: pkg/wifi/manager.go
//go:build !no_wifi

package wifi

import (
	"context"
	"fmt"
	"github.com/puzpuzpuz/xsync"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"
)

type ConnectedCallback func(ctx context.Context, device AnovaDevice)

type AnovaManager interface {
	Devices() []AnovaDevice
	Device(deviceID string) AnovaDevice
	OnDeviceConnected(callback ConnectedCallback)
	OnDeviceDisconnected(deviceID string, callback DisconnectedCallback)
	OnDeviceStateChange(deviceID string, callback StateChangeCallback)
	OnDeviceEvent(deviceID string, callback DeviceEventCallback)
	Server() AnovaServer
	Close() error
}

type manager struct {
	server                      AnovaServer
	devices                     *xsync.MapOf[string, AnovaDevice]
	deviceConnectedCallbacks    []ConnectedCallback
	deviceDisconnectedCallbacks *xsync.MapOf[string, DisconnectedCallback]
	deviceStateChangeCallbacks  *xsync.MapOf[string, StateChangeCallback]
	deviceEventCallbacks        *xsync.MapOf[string, DeviceEventCallback]
	logger                      *zap.SugaredLogger
}

func NewAnovaManager(ctx context.Context, host string, port int, logger *zap.Logger) (AnovaManager, error) {
	if logger == nil {
		logger = zap.NewNop()
	}
	logger = logger.Named("wifi_manager")

	srv, err := NewAnovaServer(ctx, host, port, logger)
	if err != nil {
		return nil, fmt.Errorf("failed to create AnovaServer: %w", err)
	}

	mgr := &manager{
		server:                      srv,
		devices:                     xsync.NewMapOf[AnovaDevice](),
		deviceDisconnectedCallbacks: xsync.NewMapOf[DisconnectedCallback](),
		deviceStateChangeCallbacks:  xsync.NewMapOf[StateChangeCallback](),
		deviceEventCallbacks:        xsync.NewMapOf[DeviceEventCallback](),
		logger:                      logger.Sugar(),
	}
	mgr.server.OnConnection(mgr.handleNewConnection)
	return mgr, nil
}

func (m *manager) Server() AnovaServer {
	return m.server
}

func (m *manager) Close() error {
	eg := errgroup.Group{}
	m.devices.Range(func(key string, dev AnovaDevice) bool {
		func(dev AnovaDevice) {
			eg.Go(func() error {
				return dev.Close()
			})
		}(dev)
		return true
	})
	err := eg.Wait()
	if err != nil {
		return err
	}
	m.devices = xsync.NewMapOf[AnovaDevice]()
	return m.server.Close()
}

func (m *manager) Devices() []AnovaDevice {
	var devices []AnovaDevice
	m.devices.Range(func(_ string, device AnovaDevice) bool {
		devices = append(devices, device)
		return true
	})
	return devices
}

func (m *manager) Device(deviceID string) AnovaDevice {
	dev, ok := m.devices.Load(deviceID)
	if !ok {
		return nil
	}
	return dev
}

func (m *manager) OnDeviceConnected(callback ConnectedCallback) {
	m.deviceConnectedCallbacks = append(m.deviceConnectedCallbacks, callback)
}

func (m *manager) OnDeviceDisconnected(deviceID string, callback DisconnectedCallback) {
	m.deviceDisconnectedCallbacks.Store(deviceID, callback)
}

func (m *manager) OnDeviceStateChange(deviceID string, callback StateChangeCallback) {
	m.deviceStateChangeCallbacks.Store(deviceID, callback)
}

func (m *manager) OnDeviceEvent(deviceID string, callback DeviceEventCallback) {
	m.deviceEventCallbacks.Store(deviceID, callback)
}

func (m *manager) handleNewConnection(ctx context.Context, conn AnovaConnection) error {
	dev, err := NewAnovaDevice(ctx, conn, m.logger.Desugar())
	if err != nil {
		return fmt.Errorf("failed to create AnovaDevice: %w", err)
	}
	deviceID := dev.IDCard()
	if deviceID == "" {
		return fmt.Errorf("device ID is empty after handshake")
	}

	if oldDevice, loaded := m.devices.LoadAndStore(deviceID, dev); loaded {
		m.logger.With("device", deviceID).Debug("Device is already connected. Closing old connection.")
		if err := oldDevice.Close(); err != nil {
			m.logger.With("device", deviceID).With("error", err).Error("Error closing old device connection")
		}
	}

	dev.SetStateChangeCallback(m.handleDeviceStateChange)
	dev.SetEventCallback(m.handleDeviceEvent)
	dev.SetHandleDisconnectCallback(m.handleDeviceDisconnection)

	m.logger.With("device", deviceID).Info("New device connected")

	for _, callback := range m.deviceConnectedCallbacks {
		callback(ctx, dev)
	}

	return nil
}

func (m *manager) handleDeviceDisconnection(ctx context.Context, deviceID string) {
	if dev, loaded := m.devices.LoadAndDelete(deviceID); loaded {
		m.logger.With("device", deviceID).Info("Device disconnected")
		if err := dev.Close(); err != nil {
			m.logger.With("device", deviceID).With("error", err).Error("Error closing device")
		}

		if callback, ok := m.deviceDisconnectedCallbacks.Load(deviceID); ok {
			callback(ctx, deviceID)
		}
		if callback, ok := m.deviceDisconnectedCallbacks.Load("*"); ok {
			callback(ctx, deviceID)
		}

		m.deviceDisconnectedCallbacks.Delete(deviceID)
		m.deviceStateChangeCallbacks.Delete(deviceID)
		m.deviceEventCallbacks.Delete(deviceID)
	}
}

func (m *manager) handleDeviceStateChange(ctx context.Context, deviceID string, state DeviceState) {
	if callback, ok := m.deviceStateChangeCallbacks.Load(deviceID); ok {
		callback(ctx, deviceID, state)
	}
	if callback, ok := m.deviceStateChangeCallbacks.Load("*"); ok {
		callback(ctx, deviceID, state)
	}
}

func (m *manager) handleDeviceEvent(ctx context.Context, deviceID string, event AnovaEvent) {
	if callback, ok := m.deviceEventCallbacks.Load(deviceID); ok {
		callback(ctx, deviceID, event)
	}
	if callback, ok := m.deviceEventCallbacks.Load("*"); ok {
		callback(ctx, deviceID, event)
	}
}

File: python/src/anova_wifi/manager.py
import asyncio
import logging
from typing import Dict, List, Callable, Coroutine, Any, Optional

from .connection import AnovaConnection
from .device import AnovaDevice, DeviceState
from .event import AnovaEvent
from .server import AnovaServer

logger = logging.getLogger(__name__)

HEARTBEAT_INTERVAL = 3  # seconds


class AnovaManager:
    server: AnovaServer
    devices: Dict[str, AnovaDevice] = {}
    _monitoring_tasks: Dict[str, asyncio.Task[None]] = {}

    device_connected_callbacks: List[Optional[Callable[[AnovaDevice], Coroutine[None, None, None]]]] = []
    device_disconnected_callbacks: Dict[str, Optional[Callable[[str], Coroutine[None, None, None]]]] = {}
    device_state_change_callbacks: Dict[str, Optional[Callable[[str, DeviceState], Coroutine[None, None, None]]]] = {}
    device_event_callbacks: Dict[str, Optional[Callable[[str, AnovaEvent], Coroutine[None, None, None]]]] = {}

    def __init__(self, host: str = "0.0.0.0", port: int = 8080):
        self.server = AnovaServer(host, port)

    async def start(self) -> None:
        """
        Start the AnovaManager
        :return:
        """
        self.server.on_connection(self._handle_new_connection)
        await self.server.start()
        logger.info(f"AsyncAnovaManager started on {self.server.host}:{self.server.port}")

    async def stop(self) -> None:
        """
        Stop the AnovaManager and close all devices
        :return:
        """
        await self._stop_all_monitoring_tasks()
        await self._close_all_devices()
        await self.server.stop()

        logger.info("AsyncAnovaManager stopped")

    async def _stop_all_monitoring_tasks(self) -> None:
        for device_id, task in self._monitoring_tasks.items():
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass
        self._monitoring_tasks.clear()

    async def _close_all_devices(self) -> None:
        for device in self.devices.values():
            await device.close()
        self.devices.clear()

    def get_devices(self) -> List[AnovaDevice]:
        """
        Get a list of all connected devices
        :return: List of AnovaDevice objects
        """
        return list(self.devices.values())

    def get_device(self, device_id: str) -> Optional[AnovaDevice]:
        """
        Get a device by its ID
        :param device_id: The device ID
        :return: AnovaDevice object or None if not found
        """
        return self.devices.get(device_id)

    def on_device_connected(self, callback: Callable[[AnovaDevice], Coroutine[Any, Any, None]]) -> int:
        """
        Register a callback for when a new device is connected
        :param callback: The callback function of the form `async def callback(device: AnovaDevice)`
        :return: The callback ID
        """
        self.device_connected_callbacks.append(callback)
        return len(self.device_connected_callbacks) - 1

    def remove_device_connected_callback(self, callback_id: int) -> None:
        """
        Remove a device connected callback
        :param callback_id: The callback ID returned by `on_device_connected`
        :return: None
        """
        self.device_connected_callbacks[callback_id] = None

    def on_device_disconnected(self, device_id: str, callback: Callable[[str], Coroutine[Any, Any, None]]) -> None:
        """
        Register a callback for when a device is disconnected
        :param device_id: The device ID (use "*" for all devices)
        :param callback: The callback function of the form `async def callback(device_id: str)`
        :return:
        """
        self.device_disconnected_callbacks[device_id] = callback

    def remove_device_disconnected_callback(self, device_id: str) -> None:
        """
        Remove a device disconnected callback
        :param device_id:
        :return:
        """
        self.device_disconnected_callbacks[device_id] = None

    def on_device_state_change(self, device_id: str,
                               callback: Callable[[str, DeviceState], Coroutine[Any, Any, None]]) -> None:
        """
        Register a callback for when a device's state changes
        :param device_id: The device ID (use "*" for all devices)
        :param callback: The callback function of the form `async def callback(device_id: str, state: DeviceState)`
        :return:
        """
        self.device_state_change_callbacks[device_id] = callback

    def remove_device_state_change_callback(self, device_id: str) -> None:
        """
        Remove a device state change callback
        :param device_id: The device ID (use "*" for all devices)
        :return:
        """
        self.device_state_change_callbacks[device_id] = None

    def on_device_event(self, device_id: str, callback: Callable[[str, AnovaEvent], Coroutine[Any, Any, None]]) -> None:
        """
        Register a callback for when a device sends an event
        :param device_id: The device ID (use "*" for all devices)
        :param callback: The callback function of the form `async def callback(device_id: str, event: AnovaEvent)`
        :return:
        """
        self.device_event_callbacks[device_id] = callback

    def remove_device_event_callback(self, device_id: str) -> None:
        """
        Remove a device event callback
        :param device_id: The device ID (use "*" for all devices)
        :return:
        """
        self.device_event_callbacks[device_id] = None

    async def _handle_new_connection(self, connection: AnovaConnection) -> None:
        device = AnovaDevice(connection)
        await device.perform_handshake()

        device_id = device.id_card
        if device_id is None:
            raise ValueError("Device ID is None after handshake")

        if device_id in self.devices:
            logger.warning(f"Device with ID {device_id} is already connected. Closing old connection.")
            await self._handle_device_disconnection(device_id)

        self.devices[device_id] = device
        device.add_state_change_callback(self._handle_device_state_change)
        device.add_event_callback(self._handle_device_event)

        self._monitoring_tasks[device_id] = asyncio.create_task(self._monitor_device(device))

        logger.info(f"New device connected: {device}")

        for callback in self.device_connected_callbacks:
            if callback:
                await callback(device)

    async def _monitor_device(self, device: AnovaDevice) -> None:
        while True:
            try:
                await device.heartbeat()
                await asyncio.sleep(HEARTBEAT_INTERVAL)
            except asyncio.CancelledError:
                break
            except Exception as e:
                if device.id_card is None:
                    logger.error(f"Device ID is None, closing connection: {e}")
                    await device.close()
                    break
                logger.error(f"Error monitoring device {device.id_card}: {e}")
                await self._handle_device_disconnection(device.id_card)
                raise

    async def _handle_device_disconnection(self, device_id: str) -> None:
        if device_id in self.devices:
            device = self.devices.pop(device_id)
            logger.info(f"Device disconnected: {device}")

            if device_id in self._monitoring_tasks:
                self._monitoring_tasks[device_id].cancel()
                del self._monitoring_tasks[device_id]

            await device.close()
            await self._handle_callback(device_id, self.device_disconnected_callbacks, device_id)

            if device_id in self.device_disconnected_callbacks:
                del self.device_disconnected_callbacks[device_id]
            if device_id in self.device_state_change_callbacks:
                del self.device_state_change_callbacks[device_id]
            if device_id in self.device_event_callbacks:
                del self.device_event_callbacks[device_id]

    async def _handle_device_state_change(self, device_id: str, state: DeviceState) -> None:
        await self._handle_callback(device_id, self.device_state_change_callbacks, device_id, state)

    async def _handle_device_event(self, device_id: str, event: AnovaEvent) -> None:
        await self._handle_callback(device_id, self.device_event_callbacks, device_id, event)

    @staticmethod
    async def _handle_callback(device_id: str, callback_dict: Dict[str, Optional[Callable]], *args,  # type: ignore
                               **kwargs) -> None:
        if "*" in callback_dict:
            cb = callback_dict["*"]
            if cb:
                await cb(*args, **kwargs)
        if device_id in callback_dict:
            cb = callback_dict[device_id]
            if cb:
                await cb(*args, **kwargs)
